### PR TITLE
Deprecate calling QueryBuilder methods with an array argument

### LINKED ISSUE #3837: Deprecate the usage of arrays in QueryBuilder methods
The usage of array arguments in `QueryBuilder::select()`, `::addSelect()`, `::groupBy()` and `::addGroupBy()` methods needs to be deprecated.

Their support was removed in https://github.com/doctrine/dbal/pull/3831.

### PR DESCRIPTION
**Pull request by BenMorel**
Fixes #3837.

### CODE REVIEW

**Comment by morozov** [comment]
Could we please annotate the corresponding arguments as variadic so that it's clear what should be used instead?

**Comment by morozov** [comment]
Otherwise, looks good.

**Comment by BenMorel** [comment]
This makes phpstan choke:

```php
$groupBy = is_array($groupBy) ? $groupBy : func_get_args();
```

> Else branch is unreachable because ternary operator condition is always true.

Any idea how to fix this?

**Comment by morozov** [comment]
> Any idea how to fix this?

Could you try adding `treatPhpDocTypesAsCertain: false` to the PHPStan configuration? See https://github.com/phpstan/phpstan/issues/2894#issuecomment-578238459. Alternatively, maybe annotate a second variadic argument instead of making the first one variadic:
```php
/**
 * @param mixed    $value
 * @param ...mixed $values
 */
function f($value/*, ...$values */)
```

**Comment by BenMorel** [comment]
> Could you try adding `treatPhpDocTypesAsCertain: false`

Didn't work. Same errors.

> Alternatively, maybe annotate a second variadic argument instead of making the first one variadic

Doesn't work if the argument is commented out:

> PHPDoc tag @param references unknown parameter: $groupBys

And if it's not commented out, back to square one:

> Call to function is_array() with string will always evaluate to false.

**Comment by morozov** [comment]
Then we’ll need to change the deprecation message:

> USING AN ARRAY ARGUMENT IS DEPRECATED. Pass each value as an individual argument.

**Comment by BenMorel** [comment]
What works is keep documenting the `array`, and the variadic as `string`. Which may make sense, after all:

```php
    /**
     * Adds a grouping expression to the query.
     *
     * USING AN ARRAY ARGUMENT IS DEPRECATED. This feature will be removed in the next major version.
     *
     * (...)
     *
     * @param string|array $groupBy     The grouping expression. USING AN ARRAY IS DEPRECATED.
     * @param string       ...$groupBys
     *
     * @return $this This QueryBuilder instance.
     */
    public function addGroupBy($groupBy, ...$groupBys)
```

**Comment by morozov** [comment]
It will introduce a BC break for extending classes (https://3v4l.org/CSa1P). Let's avoid code changes.

**Comment by BenMorel** [comment]
> It will introduce a BC break

Damn. What about this: 0af792252f4d7c62895e50a33df6500a625770b0?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 1]
```diff
@@ -451,18 +451,21 @@ public function add($sqlPartName, $sqlPart, $append = false)
...
      * <code>
      *     $qb = $conn->createQueryBuilder()
      *         ->select('u.id', 'p.id')
      *         ->from('users', 'u')
      *         ->leftJoin('u', 'phonenumbers', 'p', 'u.id = p.user_id');
      * </code>
      *
-     * @param mixed $select The selection expressions.
+     * @param string|string[]|null $select The selection expression. USING AN ARRAY OR NULL IS DEPRECATED.
+     *                                     Pass each value as an individual argument. Pass at least one value.
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: > Pass at least one value.

This is not necessary. You still will be able to call it without arguments to declare a `SELECT` query. At least it’s the current state.

**Comment by morozov** [comment]
> What about this: 0af7922?

Good with me.

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 1 above)
Comment (BenMorel) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Hmm, true. Actually I think we should force at least one value in `select()` as well; but we can keep this for another time.

**Comment by BenMorel** [comment]
Removed "Pass at least one value." and squashed.

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 1 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: I deliberately kept it like for similarity with `insert()`, `update()` and `delete()`. It just defines the query type. Definitely worth revisiting later.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @BenMorel!

**Comment by BenMorel** [comment]
@morozov Have you been able to merge to `master`?

**Comment by morozov** [comment]
Just filed https://github.com/doctrine/dbal/pull/3857. There was a ton of conflicts between `3.0.x` and `master`.

**Comment by BenMorel** [comment]
I knew this would be fun. Are you doing it as a PR to get reviews?

**Comment by morozov** [comment]
It's not that _I_ want to get a review but I want the merge commit to be built in the same environment where it would be built after the merge. Also, we have all release branches protected.

**Comment by morozov** [comment]
@BenMorel, #3857 is merged.

**Comment by BenMorel** [comment]
Good job. Thank you!

**Pull request merged**